### DISCUSSION TITLE
WEB SERVER in AP MODE using asyncio [Solved]

### DISCUSSION THREAD

**Discussion by ebolisa** [micropython/micropython#14219] [RP2040 / Pico] [answered]
Hi,
I just cannot get a web server running in ap_mode. I do appreciate some help.
TIA

```
try:
    import uasyncio as asyncio
    import usocket as socket
except ImportError:
    import asyncio
    import socket

import network

ssid = 'MicroPython-AP'
password = '123456789'

async def connect_to_ap():
    ap = network.WLAN(network.AP_IF)
    ap.active(True)
    ap.config(essid=ssid, password=password)

    while not ap.active():
        await asyncio.sleep(1)

    print('Connection successful')
    print(ap.ifconfig())

async def handle_client(reader, writer):
    request = await reader.read()
    print('Content =', request.decode('utf-8'))
    response = web_page()
    writer.write(response.encode('utf-8'))
    await writer.drain()
    writer.close()

def web_page():
    html = """<html><head><meta name="viewport" content="width=device-width, initial-scale=1"></head>
    <body><h1>Hello, World!</h1></body></html>"""
    return html

async def main():
    await connect_to_ap()

    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    # server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind(('0.0.0.0', 80))
    server.listen(5)

    print('Web server started')
    while True:
        client_socket, _ = await asyncio.sock_accept(server)
        asyncio.create_task(handle_client(client_socket, None))

try:
    asyncio.run(main())
finally:
    asyncio.new_event_loop()
```

**Comment by bulletmark** [comment]
It's much easier, and more robust and comprehensive,  to do the web server part like this: https://microdot.readthedocs.io/en/latest/intro.html#a-simple-microdot-web-server

**Reply by ebolisa** [reply]
> It's much easier, and more robust and comprehensive,  to do the web server part like this: https://microdot.readthedocs.io/en/latest/intro.html#a-simple-microdot-web-server

I'm not familiar with Microdot. However, even if it will put me on a different path, will take a look and see if I can crunch a code with it. Thank you!

**Comment by shariltumin** [comment]
I have made some changes to your code. I've tested the code on "MicroPython v1.22.2 on 2024-02-22; Generic ESP32 module with ESP32".

```py
import asyncio
import socket
import network

ssid = 'MicroPython-AP'
password = '123456789'
count = 1

async def connect_to_ap():
    ap = network.WLAN(network.AP_IF)
    ap.active(True)
    ap.config(essid=ssid, authmode=3, password=password)
    ap.ifconfig(('10.10.4.1', '255.255.255.0', '10.10.4.1', '1.1.1.1'))

    while not ap.active():
        await asyncio.sleep(1)

    print('Connection successful')
    print(ap.ifconfig())

async def handle_client(reader, writer):
    global count
    print(f'Handle #{count}')
    request = await reader.readline()
    print('Content =', request.decode('utf-8'))
    response = web_page()
    writer.write(response.encode('utf-8'))
    await writer.drain()
    writer.close()
    print('Done!')

def web_page():
    global count
    html = f"""<html><head><meta name="viewport" content="width=device-width, initial-scale=1"></head>
    <body><h1>#{count} Hello, World!</h1></body></html>"""
    count+=1
    return html

async def server():
    server = await asyncio.start_server(handle_client, '0.0.0.0', 80)
    print('Web server started')
    await server.wait_closed()

async def main():
    await connect_to_ap()
    await server()

asyncio.run(main())
```

It works, hopefully it works for you too. Sometimes we just need a simple http server without the need for a full framework like microdot.

**Reply by shariltumin** [reply]
The ```connect_to_ap()``` is for esp32 and does not work for rp2040. It works if I change it to:

```py
async def connect_to_ap():
    ap = network.WLAN(network.AP_IF)
    ap.config(essid=ssid, password=password)
    ap.active(True)

    while not ap.active():
        await asyncio.sleep(1)

    print('Connection successful')
    print(ap.ifconfig())
```

In this case the server address will be '192.168.4.1'.

**Reply by ebolisa** [reply]
Thank you for getting back to me. I just tried it on a pico and it does display my browser connections/counts on the console
```
Done!
Handle #9
Content = GET / HTTP/1.1
```
but the browsers (Chrome/Safari) refuse to connect thus not displaying the web page.
Will try on an ESP.

UPDATE:
Same results with ESP. BTW, I think the line `ap.ifconfig(('10.10.4.1', '255.255.255.0', '10.10.4.1', '1.1.1.1')) `should be omitted.

**Reply by shariltumin** [reply, marked as answer]
I need the ap.ifconfig(...) as my home router has the same IP as the default AP IP address. And it works for esp32. I need to recompile the RP2040 firmware to change the AP address, unfortunately this cannot be done from MP.

If you see ```Handle #9```, then your browser request reached the server, but somehow the response never got back to the client.

This change to web_page() will ensure that the response contains correctly formatted html.

```py
def web_page():
    global count
    page = f"""<html><head><meta name="viewport" content="width=device-width, initial-scale=1"></head>
    <body><h1>#{count} Hello, World!</h1></body></html>"""
    html = f"""HTTP/1.1 200 OK
Content-Type: text/html
Content-Length: {len(page)}
Connection: Closed

"""
    count+=1
    return html+page
```

We need to make sure that the client socket is properly closed:

```py
async def handle_client(reader, writer):
    global count
    print(f'Handle #{count}')
    request = await reader.readline()
    print('Content =', request.decode('utf-8'))
    response = web_page()
    writer.write(response.encode('utf-8'))
    await writer.drain()
    writer.close()
    await writer.wait_closed()
    print('Done!')
```
Otherwise ESP32 will stop responding after 10 requests.

**Reply by ebolisa** [reply]
Hi, 

At first, the revised code failed the same way as before but, after reviewing the issue with a fresh mind, I realized I was causing the problem by adding spaces (lines) in the html code.
Never html codes have been an issue for me but, in this case, space(s) did affect the python code.

Thank you much!!
